### PR TITLE
Unhide unique from C++, make unique partially scriptable

### PR DESCRIPTION
**Pull request by zasdfgbnm**
This PR does three things:

## Null-able integer argument in schema

~~Allow `int64_t?` in function schema,  which provide an elegant way of implementing null-able int arguments, as discussed in https://github.com/pytorch/pytorch/pull/15208#pullrequestreview-185230081~~ 

~~Originally implemented in https://github.com/pytorch/pytorch/pull/15235~~

~~Example:~~

```yaml
- func: myop(Tensor self, int64_t? dim=None) -> Tensor
  variants: function
```

~~cc: @zou3519~~

Edit: implemented in https://github.com/pytorch/pytorch/pull/15234

## Unhide unique from C++

Previously tried in https://github.com/pytorch/pytorch/pull/12064. There was a problem that C++ does not have kwarg support, which makes it confusing to know whether `unique(t, 1)` actually means `unique(t, dim=1)` or `unique(t, sorted=1)`.

Now I think I have a better idea on how to implement this: there are two ATen operators: `unique` and `unique_dim`. `unique` has the same signature as in python, and exported to both python and C++. `unique_dim` has signature `unique_dim(tensor, dim, sorted=False, return_inverse=False)`, and only exported to C++, which could be used more naturally for a C++ user.

cc: @ezyang @t-vi How does it look like this time?

## Unique is now partially scriptable

As a result of the above two, `torch.unique` can now be partially scripted. See test below:

```python
import torch

def f(a):
    b, c = torch.unique(a, return_inverse=True)
    return b.sum()

def g(a):
    b = torch.unique(a)
    return b.sum()

a = torch.rand((5, 6, 7))

traced_f = torch.jit.trace(f, a)
traced_g = torch.jit.trace(g, a)
print('traced f:', traced_f.graph)
print('traced g:', traced_g.graph)

scripted_f = torch.jit.script(f)
print('scripted f:', scripted_f.graph)

scripted_g = torch.jit.script(g)
print('scripted g:', scripted_g.graph)
```

gives

```
traced f: graph(%input : Float(5, 6, 7)) {
  %1 : bool = prim::Constant[value=0]()
  %2 : bool = prim::Constant[value=1]()
  %3 : int? = prim::None()
  %b : Float(210), %5 : Long(5, 6, 7) = aten::unique(%input, %1, %2, %3)
  %6 : Float() = aten::sum(%b)
  return (%6);
}

traced g: graph(%input : Float(5, 6, 7)) {
  %1 : bool = prim::Constant[value=0]()
  %2 : bool = prim::Constant[value=0]()
  %3 : int? = prim::None()
  %b : Float(210), %5 : Long(0) = aten::unique(%input, %1, %2, %3)
  %6 : Float() = aten::sum(%b)
  return (%6);
}

scripted f: graph(%a : Tensor) {
  %2 : bool = prim::Constant[value=0]()
  %1 : bool = prim::Constant[value=1]()
  %5 : int? = prim::None()
  %b : Tensor, %c : Tensor = aten::unique(%a, %2, %1, %5)
  %9 : Tensor = aten::sum(%b)
  return (%9);
}

Traceback (most recent call last):
  File "test.py", line 21, in <module>
    scripted_g = torch.jit.script(g)
  File "/home/gaoxiang/.virtualenvs/pt/lib/python3.7/site-packages/torch/jit/__init__.py", line 692, in script
    _jit_script_compile(mod, ast, _rcb, get_default_args(fn))
RuntimeError: 
arguments for call are not valid:
  
  for operator aten::sum(Tensor self, int[] dim, bool keepdim=<default>, *, Tensor out) -> Tensor:
  expected a value of type Tensor for argument 'self' but found (Tensor, Tensor)
  def g(a):
      b = torch.unique(a)
      return b.sum()
             ~~~~~ <--- HERE
  
  for operator aten::sum(Tensor self, int[] dim, *, int dtype, Tensor out) -> Tensor:
  expected a value of type Tensor for argument 'self' but found (Tensor, Tensor)
  def g(a):
      b = torch.unique(a)
      return b.sum()
             ~~~~~ <--- HERE
  
  for operator aten::sum(Tensor self, int[] dim, bool keepdim, *, int dtype, Tensor out) -> Tensor:
  expected a value of type Tensor for argument 'self' but found (Tensor, Tensor)
  def g(a):
      b = torch.unique(a)
      return b.sum()
             ~~~~~ <--- HERE
  
  for operator aten::sum(Tensor self) -> Tensor:
  expected a value of type Tensor for argument 'self' but found (Tensor, Tensor)
  def g(a):
      b = torch.unique(a)
      return b.sum()
             ~~~~~ <--- HERE
  
  for operator aten::sum(Tensor self, *, int dtype) -> Tensor:
  expected a value of type Tensor for argument 'self' but found (Tensor, Tensor)
  def g(a):
      b = torch.unique(a)
      return b.sum()
             ~~~~~ <--- HERE
  
  for operator aten::sum(Tensor self, int[] dim, bool keepdim=<default>) -> Tensor:
  expected a value of type Tensor for argument 'self' but found (Tensor, Tensor)
  def g(a):
      b = torch.unique(a)
      return b.sum()
             ~~~~~ <--- HERE
  
  for operator aten::sum(Tensor self, int[] dim, *, int dtype) -> Tensor:
  expected a value of type Tensor for argument 'self' but found (Tensor, Tensor)
  def g(a):
      b = torch.unique(a)
      return b.sum()
             ~~~~~ <--- HERE
  
  for operator aten::sum(Tensor self, int[] dim, bool keepdim, *, int dtype) -> Tensor:
  expected a value of type Tensor for argument 'self' but found (Tensor, Tensor)
  def g(a):
      b = torch.unique(a)
      return b.sum()
             ~~~~~ <--- HERE
for call at:
def g(a):
    b = torch.unique(a)
    return b.sum()
           ~~~~~ <--- HERE
```

### CODE REVIEW

File: torch/csrc/utils/python_arg_parser.h
Comment (wanchaol) [inline torch/csrc/utils/python_arg_parser.h]: Aha, looks like I already add the optional<int> in #15234, so you should already be able to use it :)

File: aten/src/ATen/native/cuda/Unique.cu
Comment (wanchaol) [inline aten/src/ATen/native/cuda/Unique.cu]: why removing this function? shouldn't cpu and cuda share the same logic?

File: aten/src/ATen/native/native_functions.yaml
Comment (wanchaol) [inline aten/src/ATen/native/native_functions.yaml]: you mean unique_dim is not exposed to python?

Comment (zasdfgbnm) [reply aten/src/ATen/native/native_functions.yaml]: Yes, thanks for figuring out

File: aten/src/ATen/native/cuda/Unique.cu
Comment (zasdfgbnm) [reply aten/src/ATen/native/cuda/Unique.cu]: Because `unique_dim` now just calls `unique`, so the logic here is moved to `_unique_cuda`

Comment (zasdfgbnm) [reply aten/src/ATen/native/cuda/Unique.cu]: `_unique_dim_cpu` is removed as well

**Comment by zasdfgbnm** [comment]
@wanchaol I've updated this PR, and removed what is done in https://github.com/pytorch/pytorch/pull/15234.  Currently, the diff in `test_jit.py` was added to test if JIT correctly handle optional int parameter. Now I have reverted all the `int64_t?` changes except this one. Do you think it is good to keep or delete this change?

**Comment by wanchaol** [comment]
> @wanchaol The `int64_t?` part of this PR is exactly the same as what is doing in #15234 . When I started this, #15234 was not there, so I did it myself. Now we have it, and I will revert the changes of this part from this PR.

I see, `int64_t?` part was also needed when we try to remove `python_default_init` from ATen in #15234 , so I added the support there.


> @wanchaol I've updated this PR, and removed what is done in #15234. The diff in `test_jit.py` was added to test if JIT correctly handle optional int parameter. Now I have reverted all the `int64_t?` changes except this one. Do you think it is good to keep or delete this change?

Thanks! I think it's good to keep it to ensure the unique is working end to end :), maybe we can rename the test to something like `test_script_unique_none` under TestScript since it primarily are testing the `unique` functionality. Also try explicitly pass None to it to make sure it works? you can see a example for clamp in [test_jit](https://github.com/pytorch/pytorch/blob/52699f07547366d93ec886972bd8e7da65be6927/test/test_jit.py#L4129)

File: tools/autograd/derivatives.yaml
Comment (wanchaol) [inline tools/autograd/derivatives.yaml]: is there reasons for not implementing autograd for unique?

File: torch/functional.py
Comment (wanchaol) [inline torch/functional.py]: why we don't use the same torch.unique as in torch/tensor.py?

File: tools/autograd/derivatives.yaml
Comment (zasdfgbnm) [reply tools/autograd/derivatives.yaml]: What happens if we didn't put an operator to this list? Isn't `not_implemented` the default? I'm not sure about this. 

I think the only reason for not supporting unique's autograd is, nobody need it, so nobody complain about it, so nobody go ahead and implement it...

Comment (wanchaol) [reply tools/autograd/derivatives.yaml]: sure if nobody wants it then we don't need to implement it, but I think it's good to keep the correct function signature here, in your case this might be `unique(Tensor self, bool sorted, bool return_inverse, int64_t? dim)`, and keep the `not_implemented` here because it will be used to give decent error msg to user that this function is not having autograd implemented yet.

**Comment by zasdfgbnm** [comment]
@wanchaol Just added unique back to `derivatives.yaml`

**Review by wanchaol — APPROVED** [review]
LGTM now. @ezyang @t-vi if no more comments on the C++/Python API, I will go ahead and merge this.

**Comment by wanchaol** [comment]
There looks to be some failures in ONNX path, our onnx symbolic is only handling the old code behavior as indicated [here](https://github.com/pytorch/pytorch/blob/f3a588fedebb8e4e131aa734f03e4e61142120d0/torch/onnx/symbolic.py#L975), can you also fix it?

**Comment by t-vi** [comment]
I'm ignorant of whether we have a general policy of how to deal with Python kw args in C++, and I don't really have an opinion on the specific case either...

**Comment by zou3519** [comment]
@wanchaol is it possible to add an OSS test that covers the ONNX symbolic?

**Comment by wanchaol** [comment]
@zou3519 Yes, we have oss test in [test_operators](https://github.com/pytorch/pytorch/blob/master/test/onnx/test_operators.py)

**Comment by zasdfgbnm** [comment]
@wanchaol I've modified the symbolic and added test. How does it looks this time?

**Comment by zou3519** [comment]
@zasdfgbnm @wanchaol says it looks fine.

**Comment by ezyang** [comment]
It looks like there is still something not right with the `gen_aten_op.h` integration. Internal tests report this error:

```
Exception: Test of model [redacted] failed, 
terminate called after throwing an instance of 'c10::Error'
  what():  [enforce fail at gen_aten_op.h:15009] . Attempting to run unknown ATen operator configuration: _unique-return_inverse-sorted-1
```

@wanchaol could you please take a look?

**Comment by wanchaol** [comment]
> gen_aten_op

@ezyang Sure I am looking into it

**Comment by wanchaol** [comment]
@zasdfgbnm I investigated it with our internal failures, it is basically a backward compatibility issue.  There're regression tests that indicates multiple serialized models have "_unique" as the operator name, so if the PR change the name of it, the serialized models will break, and per @jhcross, we could not automatically redeploy the binaries without retraining(or somehow updating) all of those models. 

This should be a BC change, since pytorch and c2 don't have operator versioning for serialized models, we can only wait for either operator versioning available, or we retrained all of the serialized models. As of for now, in order to unblock this, one possible solution is that we keep both names for "_unique" and "unique" available, "_unique" just need to call "unique" in this PR.

**Comment by zasdfgbnm** [comment]
@wanchaol What happened to this? I saw it merged but then reverted. Do I need to do anything?

**Comment by wanchaol** [comment]
> @wanchaol What happened to this? I saw it merged but then reverted. Do I need to do anything?

@zasdfgbnm this still cause some internal integration tests failed (it is not being observed until we landed it). I will do some investigation and get back to you with the details

**Comment by zasdfgbnm** [comment]
closed in favor of https://github.com/pytorch/pytorch/pull/17097

**Pull request closed without merging**
